File: Session_06_DynamicProgramming_Classic/multi_language_implementations/knapsack_01.java
import java.util.Scanner;

public class knapsack_01 {
    public static void solveKnapsack(int[] weights, int[] values, int n, int W) {
        int[][] dp = new int[n + 1][W + 1];
        for (int i = 0; i <= n; i++) {
            for (int w = 0; w <= W; w++) {
                if (i == 0 || w == 0) {
                    dp[i][w] = 0;
                } else if (weights[i - 1] <= w) {
                    dp[i][w] = Math.max(dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }
        System.out.println("Max Knapsack value: " + dp[n][W]);
        int w = W;
        for (int i = n; i > 0; i--) {
            if (dp[i][w] != dp[i - 1][w]) {
                System.out.println("  Selected: Item ID " + i + " (Val: " + values[i - 1] + ", Wt: " + weights[i - 1] + ")");
                w -= weights[i - 1];
            }
        }
    }

    public static void main(String[] args) {
        int[] values = {60, 100, 120};
        int[] weights = {10, 20, 30};
        solveKnapsack(weights, values, 3, 50);
        System.out.println();
        System.out.print("Press Enter to exit...");
        Scanner scanner = new Scanner(System.in);
        scanner.nextLine();
        scanner.close();
    }
}

File: Session_06_DynamicProgramming_Classic/multi_language_implementations/knapsack_01.kt
import java.util.Scanner
import kotlin.math.max

fun solveKnapsack(weights: IntArray, values: IntArray, n: Int, W: Int) {
    val dp = Array(n + 1) { IntArray(W + 1) }
    for (i in 0..n) {
        for (w in 0..W) {
            if (i == 0 || w == 0) {
                dp[i][w] = 0
            } else if (weights[i - 1] <= w) {
                dp[i][w] = max(dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]])
            } else {
                dp[i][w] = dp[i - 1][w]
            }
        }
    }
    println("Max Knapsack value: ${dp[n][W]}")
    var w = W
    for (i in n downTo 1) {
        if (dp[i][w] != dp[i - 1][w]) {
            println("  Selected: Item ID $i (Val: ${values[i - 1]}, Wt: ${weights[i - 1]})")
            w -= weights[i - 1]
        }
    }
}

fun main() {
    val values = intArrayOf(60, 100, 120)
    val weights = intArrayOf(10, 20, 30)
    solveKnapsack(weights, values, 3, 50)
    println()
    print("Press Enter to exit...")
    Scanner(System.`in`).nextLine()
}
